crates/transaction-pool/src/pool/parked.rs

```diff
@@ -73,16 +90,102 @@ impl<T: ParkedOrd> ParkedPool<T> {
         let tx = self.by_id.remove(id)?;
         self.best.remove(&tx);
 
+        // remove from by_sender, if there are no more elements in btreeset
+        if let Some(entries) = self.by_sender.get_mut(&tx.transaction.sender()) {
+            entries.remove(&EvictionEntry { id: *id, submission_id: tx.submission_id });
+            if entries.is_empty() {
+                self.by_sender.remove(&tx.transaction.sender());
+            }
+        }
+
         // keep track of size
         self.size_of -= tx.transaction.size();
 
         Some(tx.transaction.into())
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.best.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Get transactions by sender
+    pub(crate) fn get_txs_by_sender(
+        &self,
+        sender: &Address,
+    ) -> Vec<Arc<ValidPoolTransaction<T::Transaction>>> {
+        let Some(tx_ids) = self.by_sender.get(sender) else {
+            // no transactions for this sender
+            return Vec::new()
+        };
+
+        let tx_ids = tx_ids.iter().map(|entry| entry.id).collect::<HashSet<_>>();
+
+        let mut txs = Vec::new();
+        for tx_id in tx_ids {
+            let tx = self.by_id.get(&tx_id).expect("transaction exists");
+            txs.push(tx.transaction.clone().into());
+        }
+        txs
+    }
+
+    /// Returns addresses sorted by their last submission id. Addresses with older last submission
+    /// ids are first. Note that _last_ submission ids are the newest submission id for that
+    /// sender, so this sorts senders by the last time they submitted a transaction in descending
+    /// order. Senders that have least recently submitted a transaction are first.
+    ///
+    /// Similar to `Heartbeat` in Geth
+    pub(crate) fn get_senders_by_submission_id(&self) -> Vec<Address> {
+        // iterate through by_values, and get the last submission id for each sender
+        let mut senders = self
+            .by_sender
+            .iter()
+            .map(|(addr, entries)| {
+                // SAFETY: `entries` should never be empty
+                (addr, entries.last().unwrap().submission_id)
+            })
+            .collect::<Vec<_>>();
+
+        // sort s.t. senders with older submission ids are first
+        senders.sort_by(|a, b| b.1.cmp(&a.1));
+
+        senders.into_iter().map(|(addr, _)| *addr).collect()
```

Comment:
the new by_sender set seems unnecessary for this if we convert `by_id` to a BtreeMap.

because we can get the same thing if we can iterate over unique senders in order and record the last submission id of the sender range instead
and insert into something that sorts in place (BinaryHeap)